File: src/ReadNCharactersGivenRead4/Solution.java
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */

public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        int offset = 0;
        int cur = 0;
        char[] buf4 = new char[4];
        while (n > 0 && (cur = read4(buf4)) > 0){
            for(int i = 0; i < Math.min(cur, n); i++){
                buf[offset++] = buf4[i];
            }
            n -= cur;
        }
        return offset;
    }
}

File: src/ReadNCharactersGivenRead4/Solution.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Maximum number of characters to read (int)
        :rtype: The number of characters read (int)
        """
        bufPtr = 0
        offset = 0
        buf4 = [""] * 4
        while offset < n:
            bufCount = read4(buf4)
            if bufCount == 0: break
            while offset < n and bufPtr < bufCount:
                buf[offset] = buf4[bufPtr]
                offset += 1
                bufPtr += 1
            if bufPtr == bufCount: bufPtr = 0
        return offset